opslevel/resource_opslevel_tag.go

```diff
@@ -0,0 +1,162 @@
+			},
+			"key": {
+				Type:        schema.TypeString,
+				Description: "The key of the tag.",
+				ForceNew:    false,
+				Optional:    true,
+			},
+			"value": {
+				Type:        schema.TypeString,
+				Description: "The value of the tag.",
+				ForceNew:    false,
+				Optional:    true,
+			},
+			"returned_resource": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceTagCreate(d *schema.ResourceData, client *opslevel.Client) error {
+	resourceId := d.Get("resource_id").(string)
+	resourceType := opslevel.TaggableResource(d.Get("resource_type").(string))
+	resource, err := client.GetTaggableResource(resourceType, resourceId)
+	if err != nil {
+		return err
+	}
+
+	tagCreateInput := opslevel.TagCreateInput{
+		Key:   d.Get("key").(string),
+		Value: d.Get("value").(string),
+		Type:  resourceType,
+	}
+
+	if opslevel.IsID(resourceId) {
+		tagCreateInput.Id = opslevel.ID(resourceId)
+	} else {
+		tagCreateInput.Alias = resourceId
+	}
+	newTag, err := client.CreateTag(tagCreateInput)
+	if err != nil {
+		return err
+	}
+
+	d.SetId(string(newTag.Id))
+	d.Set("returned_resource", resource.ResourceId())
+	d.Set("resource_type", resourceType)
+
+	return resourceTagRead(d, client)
+}
+
+func resourceTagRead(d *schema.ResourceData, client *opslevel.Client) error {
+	resourceId := d.Get("returned_resource").(string)
+	resourceType := opslevel.TaggableResource(d.Get("resource_type").(string))
+	resource, err := client.GetTaggableResource(resourceType, resourceId)
+	if err != nil {
+		return err
+	}
+
+	id := d.Id()
+	tags, err := resource.GetTags(client, nil)
+	if err != nil {
+		return fmt.Errorf("Unable to get tags from '%s' with id '%s'", resourceType, resourceId)
+	}
+	tag, err := tags.GetTagById(*opslevel.NewID(id))
+	if err != nil || tag == nil {
+		return fmt.Errorf(
+			"Tag '%s' for type %s with id '%s' not found. %s",
+			id,
+			resource.ResourceType(),
+			resource.ResourceId(),
+			err,
+		)
+	}
+
+	if err := d.Set("resource_id", d.Get("resource_id").(string)); err != nil {
+		return err
+	}
+
+	if err := d.Set("resource_type", string(resource.ResourceType())); err != nil {
+		return err
+	}
+
+	if err := d.Set("key", tag.Key); err != nil {
+		return err
+	}
+
+	if err := d.Set("value", tag.Value); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func resourceTagUpdate(d *schema.ResourceData, client *opslevel.Client) error {
+	id := d.Id()
+
+	resourceId := d.Get("resource_id").(string)
+	resourceType := opslevel.TaggableResource(d.Get("resource_type").(string))
+	resource, err := client.GetTaggableResource(resourceType, resourceId)
+	if err != nil {
+		return err
+	}
+
+	input := opslevel.TagUpdateInput{
+		Id:    opslevel.ID(id),
+		Key:   d.Get("key").(string),
+		Value: d.Get("value").(string),
+	}
+	if _, err := client.UpdateTag(input); err != nil {
+		return err
+	}
+
+	d.Set("returned_resource", resource.ResourceId())
```

Comment:
i don't think we want to be updating this value.   Is there a case where you think the ID we asked for would be returned as a different ID that we need to update the state?  I think we can remove this.

Reply:
No, it should be the same... I'll drop this